rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,563 @@
+
+### Minor/Additive changes
+
+It's quite common to add a new field to an existing resource in order to support a new feature. All additive fields must
+be optional - adding a field which has a required and meaningful default value causes more issues than just lossy reads
+and writes. Since the versions of instances in the cluster may lag behind Auth considerably, any new fields which need
+an explicit default value to operate will cause problems for older agents which will not know about the value; this is
+particularly true for Roles when adding a new field which determines access to a resource.
+
+If a client attempts to update a resource with the new field during a rolling deployment when both the old and new
+versions of Auth are running simultaneously there is a 50/50 chance that writes and reads which contain the new field
+will be lossy since the old instance doesn't know about the new field. However, data loss is a problem outside of
+rolling Auth updates. Reads are going to be lossy unless the client version is at the same version or newer than Auth
+because they won't know about new optional fields. Writes are going to be lossy unless Auth is at the same or newer
+version than clients because it won't know that there are additional fields to persist. Until there is a solution for
+data loss under normal operation of a cluster we should accept the possible data loss during the very small time frame
+which a rolling deployment occurs.
+
+The risks of data loss from new fields may be mitigated in the future by the Cloud First Deployment strategy defined in
+RFD 134 and automatic client updates. If Cloud is running off of master the probability of a client attempting to
+read/write a field that Auth doesn't know about is less likely. Ensuring that clients are using the appropriate version
+for a cluster will prevent client drift that results in data loss. As long as the clients are not upgraded until after
+Auth then there is no opportunity for new fields to be accessed.
+
+### Changing the meaning of a field
+
+Altering or extending the capabilities of a field will result in backward compatible behavior. Imagine the following
+resource:
+
+```protobuf
+message Foo {
+  repeated Bar Bars = 1;
+}
+
+message Bar {
+  string Kind = 1;
+  string Namespace = 2;
+  string Name = 3;
+}
+```
+
+Adding a new kind of `Bar` in a newer version will force the old Auth instance to make a decision about how to handle
+unknown Bar kind. This is particularly troublesome when the resource is involved in access decisions. The only
+reasonable decision the old Auth server can make in this case is to deny access to prevent any possible bypasses as a
+result of not knowing how to process the new Bar.
+
+This is also true for extending the supported predicate functions available for any existing field. Older Auth instances
+will be unable to parse the new expression.
+
+### Alternate Options Considered
```

Comment:
For each of these alternate options, can you add a very clear and concise 1-2 sentences that explains why it is no longer under consideration?

We tend to go back and forth on a lot of these decisions so I think it would be good to capture why these were eliminated to prevent us from revisiting them too many times.